DataSetExplorer/DataSetSerializer/DataSetWithMetricsExporter.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using DataSetExplorer.DataSetBuilder.Model;
+using OfficeOpenXml;
+using OfficeOpenXml.Style;
+using RepositoryCompiler.CodeModel.CaDETModel.CodeItems;
+
+namespace DataSetExplorer.DataSetSerializer
+{
+    class DataSetWithMetricsExporter
+    {
+        private readonly string _templatePath = "../../../DataSetSerializer/Template/Dataset_Metrics_Template.xlsx";
+        private readonly string _exportPath;
+        private ExcelPackage template;
+        private ExcelWorksheet sheet;
+
+        public DataSetWithMetricsExporter(string exportPath)
+        {
+            _exportPath = exportPath;
+            ExcelPackage.LicenseContext = LicenseContext.NonCommercial;
+            template = new ExcelPackage(new FileInfo(_templatePath));
+            sheet = template.Workbook.Worksheets[0];
+        }
+
+        public void Export(List<Tuple<DataSetInstance, Dictionary<CaDETMetric, double>>> dataSetInstances,
```

Comment:
If I'm not mistaken this method can only be called once with the current design. For example, say we persist the dataset that has 1000 rows, which will populate the sheet and save it to file A. Next we call another dataset that has 500 rows that saves to file B. The B also has a 1000 rows since the second 500 are still stored.

This means that the client code must always instantiate the Exporter and then call Export and be aware of this detail, breaking encapsulation. In general, we expect our objects to be able to handle multiple calls of their public API, so you should recreate the sheet for export purposes.

Reply:
This method is called multiple times within Program.cs - it is currently called twice since we have two code smells. Each generated file has only as many rows as needed, there is no duplication or rows from previous files. The exporter is also instantiated as many times as needed (in Program.cs in ExportAnnotatedDataSet() method).